samples/nrf9160/modem_shell/src/srest/srest_shell.c

```diff
@@ -0,0 +1,310 @@
+			break;
+		case 's':
+			req_ctx.sec_tag = atoi(optarg);
+			if (req_ctx.sec_tag == 0) {
+				mosh_warn("sec_tag not an integer (> 0)");
+				return -EINVAL;
+			}
+			break;
+		case 'l':
+			response_buf_len = atoi(optarg);
+			if (response_buf_len == 0) {
+				mosh_warn("response buffer length not an integer (> 0)");
+				return -EINVAL;
+			}
+			break;
+		case 't':
+			req_ctx.timeout_ms = atoi(optarg);
+			if (req_ctx.timeout_ms == 0) {
+				mosh_warn("timeout not an integer (> 0)");
+				return -EINVAL;
+			}
+			req_ctx.timeout_ms *= 1000;
+			break;
+		case 'p':
+			req_ctx.port = atoi(optarg);
+			if (req_ctx.port == 0) {
+				mosh_warn("port not an integer (> 0)");
+				return -EINVAL;
+			}
+			break;
+		case 'b':
+			req_ctx.body = optarg;
+			len = strlen(optarg);
+			if (len > 0) {
+				req_ctx.body = k_calloc(len + 1, 1);
+				if (req_ctx.body == NULL) {
+					mosh_error("Cannot allocate memory for given body");
+					return -ENOMEM;
+				}
+				strcpy((char *)req_ctx.body, optarg);
+			}
+			break;
+		case 'H': {
+			bool room_available = false;
+
+			len = strlen(optarg);
+			if (len <= 0) {
+				mosh_error("No header given");
+				return -EINVAL;
+			}
+
+			/* Check if there are still room for additional header? */
+			for (i = 0; i < SREST_REQUEST_MAX_HEADERS; i++) {
+				if (!headers[i].in_use) {
+					room_available = true;
+					break;
+				}
+			}
+
+			if (!room_available) {
+				mosh_error("There are already max number (%d) of headers",
+					    SREST_REQUEST_MAX_HEADERS);
+				return -EINVAL;
+			}
+
+			for (i = 0; i < SREST_REQUEST_MAX_HEADERS; i++) {
+				if (!headers[i].in_use) {
+					headers[i].header_str = k_malloc(strlen(optarg) + 1);
+					if (headers[i].header_str == NULL) {
+						mosh_error(
+							"Cannot allocate memory for header %s",
+							optarg);
+						break;
+					}
+					headers[i].in_use = true;
+					headers_set = true;
+					strcpy(headers[i].header_str, optarg);
+					break;
+				}
+			}
+			break;
+		}
+		case 'v':
+			req_ctx.tls_peer_verify = atoi(optarg);
+			if (req_ctx.tls_peer_verify  < 0 || req_ctx.tls_peer_verify  > 2) {
+				mosh_error(
+					"Valid values for peer verify (%d) are 0, 1 and 2.",
+					req_ctx.tls_peer_verify);
+				return -EINVAL;
+			}
+			break;
+		case '?':
+			goto show_usage;
+		default:
+			mosh_error("Unknown option. See usage:");
+			goto show_usage;
+		}
+	}
+
+	/* Check that all mandatory args were given: */
+	if (!host_set) {
+		mosh_error("Please, give all mandatory options");
+		goto show_usage;
+	}
+
+	req_ctx.header_fields = NULL;
+	if (headers_set) {
+		/* Fill given headers to req_headers[] from headers[]: */
+		for (i = 0; i < SREST_REQUEST_MAX_HEADERS; i++) {
+			for (j = 0; j < SREST_REQUEST_MAX_HEADERS; j++) {
+				headers_set = false;
+				if (headers[j].in_use) {
+					headers_set = true;
+					headers[j].in_use = false;
+					break;
+				}
+			}
+			if (headers_set) {
+				req_headers[i] = headers[j].header_str;
```

Comment:
I'm likely missing something important but I want to raise the question anyway. What is the reason why we are using `headers[j].header_str` as an intermediate storage for the header, i.e., why we don't put it directly into `req_headers[i]` in lines 209-220?

Reply:
A good comment. I'm using intermediate storage for helping the logics, especially when multiple headers. That was my 1st write for this code and most probably this can be done in easier ways. Anyways, I'll simplify the logics based on other Tommi's comments.

Reply:
Yep, I thought this intermediate storage was coming from the first draft implementation.
I think the code would simplify a lot if you get rid of `headers` and `struct rip_header_list_item`, and use NULL to indicate a free slot in `req_headers`.

Reply:
Excellent comment. Thanks @trantanen !

Reply:
Simplified.